modules/schema/src/main/java/org/apache/ignite/internal/schema/PartialBinaryTupleMatcher.java

```diff
@@ -0,0 +1,173 @@
+
+        BinaryTupleReader tuple1 = new BinaryTuple(numElements, buffer1);
+
+        BinaryTupleReader tuple2 = isBuffer2Prefix ? new BinaryTuplePrefix(numElements, buffer2) : new BinaryTuple(numElements, buffer2);
+
+        int columnsToCompare = Math.min(tuple1.elementCount(), tuple2.elementCount());
+
+        assert columnsToCompare <= numElements;
+
+        for (int i = 0; i < columnsToCompare; i++) {
+            Readability readability = tuple1.valueReadability(i);
+
+            if (readability == Readability.NOT_READABLE) {
+                return 0;
+            }
+
+            int res = compareField(i, tuple1, tuple2, readability);
+
+            if (res != 0) {
+                return res;
+            }
+
+            if (readability == Readability.PARTIAL_READABLE) {
+                return 0;
+            }
+        }
+
+        // We use the EQUALITY FLAG to determine the outcome of the comparison operation: if the flag is set, the prefix is considered
+        // larger than the tuple and if the flag is not set, the prefix is considered smaller than the tuple. This is needed to include
+        // or exclude the scan bounds.
+        if (!isBuffer2Prefix) {
+            return 0;
+        } else {
+            return -equalityFlag(buffer2);
+        }
+    }
+
+    /**
+     * Compares two tuples by column using given column index.
+     */
+    private int compareField(int colIdx, BinaryTupleReader tuple1, BinaryTupleReader tuple2, Readability readability) {
+        assert readability != Readability.NOT_READABLE : "The field is run out of inline size and cannot be compared.";
+
+        CatalogColumnCollation collation = columnCollations.get(colIdx);
+
+        boolean tuple1HasNull = tuple1.hasNullValue(colIdx);
+        boolean tuple2HasNull = tuple2.hasNullValue(colIdx);
+
+        if (tuple1HasNull && tuple2HasNull) {
+            return 0;
+        } else if (tuple1HasNull || tuple2HasNull) {
+            return collation.nullsFirst() == tuple1HasNull ? -1 : 1;
+        }
+
+        NativeType nativeType = columnTypes.get(colIdx);
+
+        int res = readability == Readability.READABLE
+                ? compareFieldValue(nativeType.spec(), tuple1, colIdx, tuple2, colIdx)
+                : compareFieldValuePartially(nativeType.spec(), tuple1, colIdx, tuple2, colIdx);
+
+        return collation.asc() ? res : -res;
+    }
+
+    private static int compareFieldValuePartially(
+            NativeTypeSpec typeSpec,
+            BinaryTupleReader partialTuple,
+            int index1,
+            BinaryTupleReader tuple2,
+            int index2
+    ) {
+        switch (typeSpec) {
+            case BYTES: {
+                partialTuple.seek(index1);
+
+                byte[] part = partialTuple.bytesValue(partialTuple.begin(), partialTuple.byteBuffer().capacity());
+
+                byte[] cmp = ByteUtils.trimToSize(tuple2.bytesValue(index2), part.length);
```

Comment:
Can we avoid copying here ?

Reply:
Yes, I can copy only part of the array from the origin bytebuffer.

Reply:
Fixed.